cilium/CFP-42904-policy-ordering.md

```diff
@@ -0,0 +1,118 @@
+
+[PR 42784](https://github.com/cilium/cilium/pull/42784) will implement the capability to explicitly order network policy rules by introducing a new *Priority* field in the internal PolicyEntry struct. Policies with lower numeric Priority values will have higher precedence. The default priority value will be 0, representing the highest precedence. To ensure predictable ordering, we will internally assign explicit priority values to all policy rules, including those that don't natively support explicit priorities.
+
+The priority field is an *int32*, but only the first 24 bits are usable with the upper 8 bits reserved by the dataplane for mapping the ProxyPort precedence. ClusterNetworkPolicy supports a maximum of 1001 priorities (0-1000) with a maximum of 25 rules in each policy, giving us a theoretical maximum of 25025 priorities. 
+
+There will be four numeric priority ranges: one for the admin tier, one for the regular tier (k8s NP and CNP/CCNP), one for the baseline tier and one for the explicit default allow or deny.
+
+
+| Tier          | Priority range    |
+| ------------- | ----------------- |
+| Admin         | 0-25024           |
+| NetworkPolicy | 100000            |
+| Baseline      | 200000-225024     |
+| Default       | 300000            |
+
+The specific ranges are chosen to provide clear separation and ample space for future expansion:
+* The ClusterNetworkPolicy spec allows for priorities 0-1000, with up to 25 rules per policy. Reserving a range of 100,000 for the Admin and Baseline tiers provides a simple mapping (as shown below) and room for growth.
+* The NetworkPolicy and Default tiers do not have explicit priorities within their tiers, but a large range is reserved for consistency and internal use.
+* CiliumNetworkPolicy and CiliumClusterwideNetworkPolicy will be assigned to the NetworkPolicy tier. This means that Admin tier policies will always override CiliumClusterwideNetworkPolicy rules.
+* The priority numbering is purely internal to Cilium and only exists in memory within the Cilium Agent. It is thus possible to change these allocations at any time without any external impact.
+* While it is possible to split the ranges on binary boundaries, there does not seem to be a compelling reason to do so.
+* The priority mapping is applied independently for ingress and egress rules.
+* The Default tier will only contain a single "allow all" or "deny all" fallback rule.
+* The proposed priority range is well under the supported maximum (1^24 - 1).
+
+Each ClusterNetworkPolicy rule will be mapped to a Cilium priority by multiplying the CNP priority with 25 and adding the zero-based relative order of the rules within each to the priority. For example:
```

Comment:
In the next day or so, I will be merging a PR that makes priority a float. That means it can be negative or decimal. It also adds a separate ordering pass, meaning that the internal datapath ordering can be completely distinct from any user-facing API.